packages/contracts/contracts/L1/tss/TssStakingSlashing.sol

```diff
@@ -0,0 +1,373 @@
+        slashRecord[message.batchIndex][message.jailNode] = true;
+
+        require(
+            ITssGroupManager(tssGroupContract).verifySign(keccak256(_messageBytes), _sig),
+            "singer not tss group pub key"
+        );
+
+        // slash tokens
+        slash(message);
+        emit Slashing(message.jailNode, message.slashType);
+    }
+
+    /**
+     * @notice slash the staker and distribute rewards to voters
+     * @param message the message about the slash infos
+     */
+    function slash(SlashMsg memory message) internal {
+        // slashing params check
+        for (uint256 i = 0; i < 2; i++) {
+            require(slashAmount[i] > 0, "have not set the slash amount");
+            require(exIncome[i] > 0, "have not set the extra income amount");
+        }
+        bytes memory jailNodePubKey = deposits[message.jailNode].pubKey;
+        if (message.slashType == SlashType.uptime) {
+            // jail and transfer deposits
+            ITssGroupManager(tssGroupContract).memberJail(jailNodePubKey);
+            transformDeposit(message.jailNode, 0, message.tssNodes);
+        } else if (message.slashType == SlashType.animus) {
+            // remove the member and transfer deposits
+            ITssGroupManager(tssGroupContract).removeMember(jailNodePubKey);
+            transformDeposit(message.jailNode, 1, message.tssNodes);
+        } else {
+            require(false, "err type for slashing");
+        }
+    }
+
+    /**
+     * @notice distribute rewards to voters
+     * @param deduction address of the punished
+     * @param slashType the type to punished
+     * @param tssNodes participants other than the initiator
+     */
+    function transformDeposit(
+        address deduction,
+        uint256 slashType,
+        address[] memory tssNodes
+    ) internal {
+        uint256 deductedAmount;
+        uint256 totalTransfer;
+        uint256 extraAmount;
+        uint256 remainder;
+        uint256 gain;
+
+        if (deposits[deduction].amount > slashAmount[slashType]) {
+            // deposit > slashAmount, deduct slashAmount then
+            // distribute additional tokens for the sender
+            deductedAmount = slashAmount[slashType];
+            deposits[deduction].amount -= slashAmount[slashType];
+            extraAmount = slashAmount[slashType] - exIncome[slashType];
+        } else if (deposits[deduction].amount > exIncome[slashType]) {
+            // exIncome < deposit <= slashAmount, deduct all token then
+            // distribute additional tokens for the sender
+            deductedAmount = deposits[deduction].amount;
+            deposits[deduction].amount = 0;
+            extraAmount = deductedAmount - exIncome[slashType];
+        } else if (deposits[deduction].amount > 0) {
+            // 0 < deposit <= exIncome, deduct all token
+            deductedAmount = deposits[deduction].amount;
+            deposits[deduction].amount = 0;
+            extraAmount = deductedAmount;
```

Comment:
so in this case, the sender will not get the any reward?

Reply:
In my consideration the sender is also included in the tssNodes.In this case, the remainder will send to the sender.